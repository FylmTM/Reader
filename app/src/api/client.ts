type Response<R> =
  | { status: "ok"; payload: R }
  | { status: "error"; message: string };

function post<T, R>(path: string, data: T): Promise<R> {
  return fetch(path, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

function get<R>(path: string): Promise<R> {
  return fetch(path, { method: "GET" })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

export default {
  get,
  post,
};
