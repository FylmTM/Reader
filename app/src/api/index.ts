import { CategoriesWithFeeds, Post, User } from "../domain";

const DELAY = 200;

function getCurrentUser(): Promise<User> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({
        id: 1,
        username: "me"
      });
    }, DELAY);
  });
}

function login(): Promise<User> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({
        id: 1,
        username: "me"
      });
    }, DELAY);
  });
}

function logout(): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

function getCategoriesWithFeeds(): Promise<CategoriesWithFeeds> {
  return new Promise<CategoriesWithFeeds>((resolve, reject) => {
    setTimeout(() => {
      resolve([
        {
          category: {
            id: 1,
            name: "General"
          },
          feeds: [
            {
              id: 1,
              kind: "RSS",
              title: "Hacker News",
              link: "https://news.ycombinator.com/",
              feed: "https://news.ycombinator.com/rss"
            },
            {
              id: 2,
              kind: "RSS",
              title: "vc.ru",
              link: "https://vc.ru/",
              feed: "https://vc.ru/rss/all"
            }
          ]
        },
        {
          category: {
            id: 2,
            name: "Gaming - Best Gaming Category in the whole World"
          },
          feeds: [
            {
              id: 3,
              kind: "RSS",
              title: "StopGame - Best Gaming Website in the World",
              link: "https://stopgame.ru",
              feed: "https://rss.stopgame.ru/rss_all.xml"
            }
          ]
        }
      ]);
    }, DELAY);
  });
}

function randomInt(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getPosts(): Promise<Array<Post>> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(
        Array(15)
          .fill(".")
          .map((_, i) => {
            return {
              id: i + 1,
              is_read: false,
              is_read_later: false,
              feed: {
                id: 1,
                category_id: 1,
                title: "Hacker News"
              },
              link: "http://example.com",
              title:
                randomInt(0, 1) === 1
                  ? "World is finally living on mars."
                  : "World is falling down to the works of ancient civilization who is going to do a revolution.",
              summary:
                randomInt(0, 1) === 1
                  ? Array(randomInt(1, 100))
                      .fill("lorem")
                      .join(" ")
                  : undefined,
              date: `${randomInt(1, 10)}h`,
              media:
                randomInt(0, 1) === 1
                  ? {
                      type: "image/jpeg",
                      // link: "http://localhost:3000/android-chrome-512x512.png"
                      link:
                        "https://stopga.me/images/video/2019/12/07/M8MFv_nN.jpg"
                    }
                  : undefined,
              comments_link:
                randomInt(0, 1) === 1
                  ? "http://example.com/#comments"
                  : undefined
            };
          })
      );
    }, DELAY);
  });
}

function markAsRead(postId: number, isRead: Boolean): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

function markAsReadLater(postId: number, isRead: Boolean): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

export default {
  getCurrentUser,
  login,
  logout,
  getCategoriesWithFeeds,
  getPosts,
  markAsRead,
  markAsReadLater
};
