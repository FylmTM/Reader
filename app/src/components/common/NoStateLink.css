span.link:hover {
    text-decoration: underline;
}
