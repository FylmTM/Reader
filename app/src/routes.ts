export const ALL_ROUTE = "/";

export const CATEGORY_ROUTE = "/category/:categoryId";
export const CATEGORY_POST_ROUTE = "/category/:categoryId/post/:postId";
export const FEED_ROUTE = "/feed/:feedId";
export const FEED_POST_ROUTE = "/feed/:feedId/post/:postId";
