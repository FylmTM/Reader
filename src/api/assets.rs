use std::borrow::Cow;
use std::ffi::OsStr;
use std::io::Cursor;
use std::path::{Path, PathBuf};

use rocket::http::ContentType;
use rocket::request::FromRequest;
use rocket::response;
use rocket::{Outcome, Request};

use crate::error::Error as E;

#[derive(RustEmbed)]
#[folder = "static"]
struct Assets;

const INDEX: &str = "index.html";

pub struct Asset {
    content: Cow<'static, [u8]>,
    content_type: Option<ContentType>,
}

impl<'a, 'r> FromRequest<'a, 'r> for Asset {
    type Error = E;

    fn from_request(request: &'a Request<'r>) -> rocket::request::Outcome<Self, E> {
        let path_raw = request.uri().path();

        // Do not try to serve assets for API.
        if path_raw.starts_with("/api") {
            return Outcome::Forward(());
        }

        let path = if path_raw == "/" {
            INDEX
        } else {
            if Path::new(path_raw).extension().is_none() {
                // Fallback all URL's that does not possibly represent file to index.html.
                // Required to support HTML5 history API for SPA application.
                INDEX
            } else {
                // We need to remove first '/' from url path, because assets paths are relative.
                &path_raw[1..]
            }
        };

        let file: Option<Cow<'static, [u8]>> = Assets::get(&path);
        let file_path = Path::new(path);
        let extension = file_path
            .extension()
            .and_then(OsStr::to_str)
            .expect("Asset extension must be present.");
        let content_type: Option<ContentType> = ContentType::from_extension(extension);

        file.map_or_else(
            || {
                debug!("Asset does not exist, forward request path={}.", path);
                Outcome::Forward(())
            },
            |content| {
                debug!("Serve asset path={}.", path);
                Outcome::Success(Asset {
                    content,
                    content_type,
                })
            },
        )
    }
}

#[get("/", rank = 98)]
pub fn index<'r>(asset: Asset) -> response::Result<'r> {
    get_asset(asset)
}

#[get("/<_file..>", rank = 99)]
pub fn assets<'r>(asset: Asset, _file: PathBuf) -> response::Result<'r> {
    get_asset(asset)
}

fn get_asset<'r>(asset: Asset) -> response::Result<'r> {
    match asset.content_type {
        Some(content_type) => response::Response::build()
            .header(content_type)
            .sized_body(Cursor::new(asset.content))
            .ok(),
        None => response::Response::build()
            .sized_body(Cursor::new(asset.content))
            .ok(),
    }
}
