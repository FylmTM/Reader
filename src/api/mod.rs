use crate::error::Error;
use rocket_contrib::json::Json;

pub mod authentication;
pub mod catchers;
pub mod endpoints;

#[derive(Serialize, Deserialize)]
pub struct ApiResponse<T> {
    status: String,
    payload: T,
}

impl<T> ApiResponse<T> {
    fn with(payload: T) -> ApiResponse<T> {
        ApiResponse {
            status: "ok".to_string(),
            payload,
        }
    }
}

type Response<T> = Result<Json<ApiResponse<T>>, Error>;

fn ok<T>(data: T) -> Response<T> {
    Ok(Json(ApiResponse::with(data)))
}
