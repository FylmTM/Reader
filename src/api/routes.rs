use std::ops::Add;

use rocket::http::{Cookie, Cookies};
use rocket_contrib::json::Json;

use crate::api::{ok, Response};
use crate::db::{self, Queries};

#[derive(Serialize, Deserialize)]
pub struct ApiKey {
    pub api_key: String,
}

#[post("/api/v1/authenticate", data = "<api_key>")]
pub fn authenticate(
    api_key: Json<ApiKey>,
    conn: db::PoolConnection,
    mut cookies: Cookies,
) -> Response<db::User> {
    let api_key = api_key.0.api_key;
    let user = conn.find_user_by_api_key(&api_key)?;
    cookies.add(
        Cookie::build("api_key", api_key)
            .path("/")
            .http_only(true)
            .expires(time::now().add(time::Duration::days(365)))
            .finish(),
    );

    ok(user)
}

#[get("/api/v1/current_user")]
pub fn current_user(user: db::User) -> Response<db::User> {
    ok(user)
}
