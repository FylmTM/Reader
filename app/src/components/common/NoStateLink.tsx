import React, { FC, useCallback } from "react";
import "./NoStateLink.css";

type NativeAnchorElement = React.DetailedHTMLProps<
  React.AnchorHTMLAttributes<HTMLAnchorElement>,
  HTMLAnchorElement
> & {
  onClickHref?: string;
};

export function navigate(href: string | undefined) {
  if (href != null) {
    // eslint-disable-next-line no-restricted-globals
    history.pushState({}, "", href);
  }
}

export const NoStateLink: FC<NativeAnchorElement> = function NoStateLink({
  href,
  onClickHref,
  onClick,
  children,
  ...props
}) {
  const handleClick = useCallback(
    event => {
      // ignores the navigation when clicked using right mouse button or
      // by holding a special modifier key: ctrl, command, win, alt, shift
      if (
        event.ctrlKey ||
        event.metaKey ||
        event.altKey ||
        event.shiftKey ||
        event.button !== 0
      ) {
        return;
      }

      event.preventDefault();
      navigate(onClickHref || href);
      onClick && onClick(event);
    },
    [href, onClick, onClickHref]
  );

  return (
    <a href={href} onClick={handleClick} {...props}>
      {children}
    </a>
  );
};

export const NoStateSpanLink: FC<NativeAnchorElement> = function NoStateSpanLink({
  href,
  onClick,
  children,
  className,
  ...props
}) {
  const handleClick = useCallback(
    event => {
      // ignores the navigation when clicked using right mouse button or
      // by holding a special modifier key: ctrl, command, win, alt, shift
      if (
        event.ctrlKey ||
        event.metaKey ||
        event.altKey ||
        event.shiftKey ||
        event.button !== 0
      ) {
        return;
      }

      event.preventDefault();
      navigate(href);
      onClick && onClick(event);
    },
    [href, onClick]
  );

  return (
    <span onClick={handleClick} className={`link ${className}`} {...props}>
      {children}
    </span>
  );
};
