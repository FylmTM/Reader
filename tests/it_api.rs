#![feature(proc_macro_hygiene, decl_macro)]
#[macro_use]
extern crate rocket;

#[macro_use]
extern crate insta;

use common::client::{self, RequestOperations, ResponseOperations};
use reader::api::routes::*;

pub mod common;

#[test]
fn test_auth_login_failed() {
    let client = client::get();
    let mut response = client
        .post("/api/v1/auth/login")
        .json(ApiKey {
            api_key: "invalid".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity())
}

#[test]
fn test_auth_login_success() {
    let client = client::get();
    let mut response = client
        .post("/api/v1/auth/login")
        .json(ApiKey {
            api_key: "api_key".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity());
    assert_eq!(response.cookies().len(), 1);
}

#[test]
fn test_auth_logout_success() {
    let client = client::authenticated();

    let mut response = client.post("/api/v1/auth/logout").dispatch();
    assert_eq!(response.entity(), client::api_ok(()));
    assert_eq!(response.cookies().len(), 1);

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user_failed_if_not_authenticated() {
    let client = client::get();

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_categories_with_feeds() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/categories_with_feeds").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_all() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_read_later() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?is_read_later=true").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_category() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?category_id=1").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_feed() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?feed_id=1").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_post_content() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts/1/content").dispatch();
    assert_json_snapshot!(response.json_entity());
}
