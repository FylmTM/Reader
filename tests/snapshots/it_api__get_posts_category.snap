---
source: tests/it_api.rs
expression: response.json_entity()
---
[
  "200 OK",
  {
    "payload": [
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21746739",
        "date": "2019-12-09T21:22:57Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 30,
        "is_read": false,
        "is_read_later": false,
        "link": "https://en.wikipedia.org/wiki/Transputer",
        "media_link": null,
        "summary": "Comments",
        "title": "Transputer"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21745731",
        "date": "2019-12-09T19:46:52Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 29,
        "is_read": false,
        "is_read_later": false,
        "link": "https://blog.cynthia.re/post/nintendo-switch-ethernet-switch",
        "media_link": null,
        "summary": "Comments",
        "title": "The Nintendo Switch Switch"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21747217",
        "date": "2019-12-09T22:14:34Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 28,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.theinformation.com/articles/apples-ad-targeting-crackdown-shakes-up-ad-market?pu=hackernews4qs3ac&utm_source=hackernews&utm_medium=unlocka",
        "media_link": null,
        "summary": "Comments",
        "title": "Apple’s Ad-Targeting Crackdown Shakes Up Ad Market"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743424",
        "date": "2019-12-09T15:30:35Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 27,
        "is_read": false,
        "is_read_later": false,
        "link": "https://randomascii.wordpress.com/2019/12/08/on2-again-now-in-wmi/",
        "media_link": null,
        "summary": "Comments",
        "title": "O(n^2), again, now in Windows Management Instrumentation"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21738802",
        "date": "2019-12-08T22:20:15Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 26,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.evanjones.ca/ordered-vs-unordered-indexes.html",
        "media_link": null,
        "summary": "Comments",
        "title": "Why databases use ordered indexes but programming uses hash tables"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21747446",
        "date": "2019-12-09T22:41:20Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 25,
        "is_read": false,
        "is_read_later": false,
        "link": "https://hbr.org/2017/09/the-grocery-industry-confronts-a-new-problem-only-10-of-americans-love-cooking",
        "media_link": null,
        "summary": "Comments",
        "title": "Grocery Industry Confronts a Problem: Only 10% of Americans Love Cooking (2017)"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21742739",
        "date": "2019-12-09T14:11:22Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 24,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.sametab.com/blog/operations-and-internal-communication-strategies-for-effective-ceos",
        "media_link": null,
        "summary": "Comments",
        "title": "Operations for Effective CEOs"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21744274",
        "date": "2019-12-09T17:10:11Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 23,
        "is_read": false,
        "is_read_later": false,
        "link": "https://signal.org/blog/signal-private-group-system/",
        "media_link": null,
        "summary": "Comments",
        "title": "Technology Preview: Signal Private Group System"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21737697",
        "date": "2019-12-08T19:31:16Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 22,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.nytimes.com/2019/12/08/obituaries/caroll-spinney-dead.html",
        "media_link": null,
        "summary": "Comments",
        "title": "Caroll Spinney has died"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21747175",
        "date": "2019-12-09T22:10:43Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 21,
        "is_read": false,
        "is_read_later": false,
        "link": "https://blog.letsenhance.io/all/2019/07/09/computational-photographyfrom-selfies-to-black-holes/",
        "media_link": null,
        "summary": "Comments",
        "title": "Computational photography from selfies to black holes"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743302",
        "date": "2019-12-09T15:17:05Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 20,
        "is_read": false,
        "is_read_later": false,
        "link": "https://news.ycombinator.com/item?id=21743302",
        "media_link": null,
        "summary": "Comments",
        "title": "Ask HN: We are shutting down our startup, I get our code. What now?"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743847",
        "date": "2019-12-09T16:20:49Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 19,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.quantamagazine.org/how-simple-rules-bootstrap-the-laws-of-physics-20191209/",
        "media_link": null,
        "summary": "Comments",
        "title": "Simple Rules ‘Bootstrap’ the Laws of Physics"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21729056",
        "date": "2019-12-07T07:51:57Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 18,
        "is_read": false,
        "is_read_later": false,
        "link": "http://www.2arms1head.com/",
        "media_link": null,
        "summary": "Comments",
        "title": "Two Arms and a Head (2011)"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21747710",
        "date": "2019-12-09T23:13:16Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 17,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.bloomberg.com/news/articles/2019-12-09/vietnam-s-richest-man-bets-2-billion-to-sell-cars-to-americans",
        "media_link": null,
        "summary": "Comments",
        "title": "Vietnam’s Richest Man Bets $2B to Sell Cars to Americans"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21748589",
        "date": "2019-12-10T01:03:09Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 16,
        "is_read": false,
        "is_read_later": false,
        "link": "https://effectiviology.com/jumping-to-conclusions/",
        "media_link": null,
        "summary": "Comments",
        "title": "Jumping to Conclusions: When People Decide Based on Insufficient Information"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743696",
        "date": "2019-12-09T15:59:55Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 15,
        "is_read": false,
        "is_read_later": false,
        "link": "https://web.archive.org/web/20030812081713/http://klausler.com/cnotes.txt",
        "media_link": null,
        "summary": "Comments",
        "title": "Some things every C programmer should know about C (2002)"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21729597",
        "date": "2019-12-07T10:55:30Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 14,
        "is_read": false,
        "is_read_later": false,
        "link": "https://emacsair.me/2017/09/01/magit-walk-through/",
        "media_link": null,
        "summary": "Comments",
        "title": "A walk through the Magit interface"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743100",
        "date": "2019-12-09T14:53:02Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 13,
        "is_read": false,
        "is_read_later": false,
        "link": "https://queue.acm.org/detail.cfm?id=3341158",
        "media_link": null,
        "summary": "Comments",
        "title": "Must and Must Not: On writing documentation"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21745704",
        "date": "2019-12-09T19:43:30Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 12,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.wsj.com/articles/american-factories-demand-white-collar-education-for-blue-collar-work-11575907185",
        "media_link": null,
        "summary": "Comments",
        "title": "College-educated workers are taking over the American factory floor"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21742110",
        "date": "2019-12-09T12:45:13Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 11,
        "is_read": false,
        "is_read_later": false,
        "link": "https://johnaugust.com/2018/missing-movies",
        "media_link": null,
        "summary": "Comments",
        "title": "Missing Movies (2018)"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21742731",
        "date": "2019-12-09T14:10:09Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 10,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.nytimes.com/2019/12/09/business/paul-a-volcker-dead.html",
        "media_link": null,
        "summary": "Comments",
        "title": "Paul Volcker has died"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21742702",
        "date": "2019-12-09T14:05:49Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 9,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.bloomberg.com/news/articles/2019-12-08/repo-blowup-was-fueled-by-big-banks-and-hedge-funds-bis-says",
        "media_link": null,
        "summary": "Comments",
        "title": "Repo Blowup Was Fueled by Big Banks and Hedge Funds, BIS Says"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21742978",
        "date": "2019-12-09T14:38:39Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 8,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.jefftk.com/p/long-bets-by-confidence-level",
        "media_link": null,
        "summary": "Comments",
        "title": "Long Bets by Confidence Level"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743513",
        "date": "2019-12-09T15:41:11Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 7,
        "is_read": false,
        "is_read_later": false,
        "link": "https://jamesmunns.com/blog/fmt-unreasonably-expensive/",
        "media_link": null,
        "summary": "Comments",
        "title": "Formatting Is Unreasonably Expensive for Embedded Rust"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21742222",
        "date": "2019-12-09T13:06:35Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 6,
        "is_read": false,
        "is_read_later": false,
        "link": "https://github.com/infostreams/db",
        "media_link": null,
        "summary": "Comments",
        "title": "Show HN: Version Control for Databases"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21744785",
        "date": "2019-12-09T18:00:56Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 5,
        "is_read": false,
        "is_read_later": false,
        "link": "http://scholar.harvard.edu/files/mickens/files/towashitallaway.pdf",
        "media_link": null,
        "summary": "Comments",
        "title": "To Wash It All Away (2014) [pdf]"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21745124",
        "date": "2019-12-09T18:41:48Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 4,
        "is_read": false,
        "is_read_later": false,
        "link": "https://www.theatlantic.com/ideas/archive/2019/11/why-child-care-so-expensive/602599",
        "media_link": null,
        "summary": "Comments",
        "title": "Why Child Care Is So Expensive"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21744772",
        "date": "2019-12-09T17:59:35Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 3,
        "is_read": false,
        "is_read_later": true,
        "link": "https://fivebooks.com/best-books/science-2019-barbara-kiser/",
        "media_link": null,
        "summary": "Comments",
        "title": "Science Books of 2019"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21743129",
        "date": "2019-12-09T14:56:35Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 2,
        "is_read": false,
        "is_read_later": true,
        "link": "https://www.wsj.com/articles/five-cities-account-for-vast-majority-of-growth-in-tech-jobs-study-finds-11575867660",
        "media_link": null,
        "summary": "Comments",
        "title": "Five cities account for vast majority of growth in U.S. tech jobs: study"
      },
      {
        "category_id": 1,
        "comments_link": "https://news.ycombinator.com/item?id=21744688",
        "date": "2019-12-09T17:50:17Z",
        "feed_id": 1,
        "feed_title": "Hacker News",
        "id": 1,
        "is_read": false,
        "is_read_later": true,
        "link": "https://savo.la/introduction-to-gate.html",
        "media_link": null,
        "summary": "Comments",
        "title": "Introduction to Gate"
      }
    ],
    "status": "ok"
  }
]
